---
title: "The Hitchhiker's Guide to scRNA-seq"
author: ""
date: "`r format(Sys.Date(), '%d/%m/%Y')`"
output: 
   html_document:
      toc: true 
      toc_float: true
      theme: united
      code_folding: hide
---

<br>

<br>

---

<br>

<br>

# Notebook

<br>

```{r setup, include=FALSE}
## Set up configuration for R chunks
knitr::opts_chunk$set(
	echo = TRUE,
	fig.align = "center",
	message = FALSE,
	warning = FALSE
)
```

```{r packages, message=FALSE, warning=FALSE}
## Import packages
library("dplyr") 
library("Seurat") # scRNA-seq analysis
library("patchwork")
if (!"SeuratWrappers" %in% installed.packages()) remotes::install_github("satijalab/seurat-wrappers", "seurat5")
library("SeuratWrappers")
if (!"batchelor" %in% installed.packages()) BiocManager::install("batchelor")
if ((!"matrixStats" %in% installed.packages()) | (packageVersion("matrixStats")>"1.1.0")) {
  remotes::install_version("matrixStats", version="1.1.0")
} 
if (!"ComplexHeatmap" %in% installed.packages()) remotes::install_github("jokergoo/ComplexHeatmap")
if (!"zellkonverter" %in% installed.packages()) BiocManager::install("zellkonverter")
```

<br>

<br>

---

<br>

<br>

## (1) Import datasets 

_(7 min)_

<br>

```{r import data}
# Import data
data.dir <- "../data"
seu <- readRDS(file = file.path(data.dir, "pbmc3k_panc8.rds"))
```

<br>

<br>

---

<br>

<br>

## (2) Assess batch effect

_(7 min)_

<br>

Some simple analyses commonly done to assess batch effects and integration requirements are:

   1. Joint data analysis: 
   
      1.1. Data projection onto a lower dimensional space, e.g., PCA and UMAP
      
      1.2. Look into expected cell type marker genes 
   
   2. Independent sample analysis: 
   
      2.1. Manual cell annotation: cluster marker genes comparison in between sample clusters
      
      2.2. Automatic cell annotation: cell/cluster annotation against a representative reference
      
<br>

### Joint dimred

<br>
      
```{r joint dimred analysis}
## Joint analysis

# Standard Seurat upstream workflow
seu <- NormalizeData(seu)
seu <- FindVariableFeatures(seu)
seu <- ScaleData(seu)
seu <- RunPCA(seu)
seu <- RunUMAP(seu, dims = 1:30, reduction = "pca", reduction.name = "umap.unintegrated")
```

```{r plot joint dimreds, fig.width=10, fig.height=4}
## Plot jointly dimreds
pca.unint <- DimPlot(seu, reduction = "pca", group.by = "batch")
umap.unint <- DimPlot(seu, reduction = "umap.unintegrated", group.by = "batch")
pca.unint + umap.unint
```

<br>

### Celltype markers

<br>

```{r joint celltype markers}
## Joint celltype markers

# List of PBMC and (some) pancreatic cell markers
markers.plot <- list(
  # "pbmc" = c("CD3D", "CREM", "HSPH1", "SELL", "GIMAP5", "CACYBP", "GNLY", "NKG7", "CCL5",
  #            "CD8A", "MS4A1", "CD79A", "MIR155HG", "NME1", "FCGR3A", "VMO1", "CCL2", "S100A9", 
  #            "HLA-DQA1", "GPR183", "PPBP", "GNG11", "HBA2", "HBB", "TSPAN13", "IL3RA", "IGJ", 
  #            "PRSS57"), 
  "pbmc" = c("CD3D", "NKG7", "CD8A", "MS4A1", "CD79A", "FCGR3A"), 
  "pancreas" = c("REG1A", "PPY", "SST", "GHRL", "VWF", "SOX10")
)

# Plot
pbmc.markers.unint.plot <- FeaturePlot(seu, features = markers.plot$pbmc, split.by = "batch", 
                                       max.cutoff = 3, cols = c("grey", "red"), 
                                       reduction = "umap.unintegrated", ncol = 4)
pancreas.markers.unint.plot <- FeaturePlot(seu, features = markers.plot$pancreas, split.by = "batch", 
                                           max.cutoff = 3, cols = c("grey", "red"), 
                                           reduction = "umap.unintegrated", ncol = 4)
```

```{r plot joint celltype markers, fig.width=8, fig.height=18}
## Plot jointly celltype markers

# Print 
pbmc.markers.unint.plot
pancreas.markers.unint.plot
```

<br>

### Manual cell annotation

<br>

```{r independent sample analysis}
## Independent sample analysis

# Split Seurat object into two batch on 'batch' label identity
seu.list <- SplitObject(object = seu, split.by = "batch")

# Standard Seurat upstream workflow
seu.list <- lapply(X = seu.list, FUN = function(x) {
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x)
  x <- ScaleData(x)
  x <- RunPCA(x)
  x <- FindNeighbors(x, dims = 1:15, reduction = "pca")
  x <- FindClusters(x, resolution = 0.8, cluster.name = "unintegrated_clusters")
  x <- RunUMAP(x, dims = 1:15, reduction = "pca", reduction.name = "umap.unintegrated")
})
```

```{r plot independent sample analysis clusters, fig.width=10, fig.height=4}
## Plot independent sample analysis clusters
umap.ind.samp.unint <- lapply(X = seu.list, FUN = function(x) {
  DimPlot(x, reduction = "umap.unintegrated", group.by = "unintegrated_clusters")
})
umap.ind.samp.unint$pbmc + umap.ind.samp.unint$pancreas
```

```{r independent sample analysis - dge}
## Independent sample analysis: DGE

# Differential gene expression analysis per cluster 
dge.markers.unint <- lapply(X = seu.list, FUN = function(x) {
  FindAllMarkers(object = x, assay = "RNA", slot = "data", 
                 logfc.threshold = 0.25, min.pct = 0.25, 
                 min.cells.feature = 10, only.pos = TRUE)
})

# Pick the top50 upregulated genes per cluster based on log2FC
top50.up.cluster <- lapply(X = dge.markers.unint, FUN = function(x) {
  x %>% 
    filter(p_val_adj<0.05) %>% 
    group_by(cluster) %>% 
    arrange(desc(avg_log2FC)) %>% 
    slice_head(n=50)  %>% 
    split(., .$cluster)
}) 

# Check intersection of top50 marker genes between clusters across batches
shared.genes <- list()
for (i in names(top50.up.cluster$pbmc)) {
  for (ii in names(top50.up.cluster$pancreas)) {
    shared.genes[[paste0("pbmc", i)]][[paste0("pancreas", ii)]] <- intersect(top50.up.cluster$pbmc[[i]]$gene, 
                                                                             top50.up.cluster$pancreas[[ii]]$gene)
  }
}
# Table with number of genes shared between pbmc vs pancreas clusters for the top50 upregulated genes per cluster
shared.genes.table <- as.data.frame(
  lapply(X = shared.genes, FUN = function(x) {
    unlist(lapply(x, length))
  })
)

```

```{r independent sample analysis - dge (table)}
## Print table
knitr::kable(shared.genes.table)
```

```{r independent sample analysis - dge (plot)}
## Plot independent sample analysis clusters
ComplexHeatmap::Heatmap(matrix = as.matrix(shared.genes.table), name = "Shared gene no.", 
                        cluster_columns = FALSE, cluster_rows = FALSE)
```

<br>

### Automatic cell annotation

<br>

```{r automatic cell annotation}
## Automatic cell annotation
# sce <- as.SingleCellExperiment(seu)
# zellkonverter::writeH5AD(sce, file = file.path("../results", "pbmc3k_pancreas.h5ad"))
```

<br>

<br>

---

<br>

<br>

## (3) Integrate datasets

_(7 min)_

<br>

```{r integration}
## Perform integration

# Split layers for integration
seu[["RNA"]] <- split(x = seu[["RNA"]], f = seu$batch)

# Standard workflow
seu <- NormalizeData(seu)
seu <- FindVariableFeatures(seu)
seu <- ScaleData(seu)
seu <- RunPCA(seu)

# Integrate layers
int.methods <- c("CCA" = "CCAIntegration", "RPCA" = "RPCAIntegration", 
                 "Harmony" = "HarmonyIntegration", "FastMNN" = "FastMNNIntegration", 
                 "scVI" = "scVIIntegration")

for (m in names(int.methods)[1:4]) {
  cat("\nRunning integration method", m, "...\n")
  int.dimred <- paste0("integrated.", m)
  umap.dimred <- paste0("umap.", m)
  # Integration
  if (m=="RPCA") {
      seu <- IntegrateLayers(object = seu, method = get(eval(substitute(int.methods[m]))), 
                         orig.reduction = "pca", 
                         new.reduction = int.dimred,
                         k.weight = 50, # otherwise it aborts 
                         verbose = TRUE)
  } else if (m=="scVI") {
          seu <- IntegrateLayers(object = seu, method = get(eval(substitute(int.methods[m]))), 
                         orig.reduction = "pca", 
                         new.reduction = int.dimred,
                         conda_env = "/home/aggs/miniconda3/envs/scvi-env", # substitute this by your installation 
                         verbose = TRUE)
  } else {
      seu <- IntegrateLayers(object = seu, method = get(eval(substitute(int.methods[m]))), 
                         orig.reduction = "pca", 
                         new.reduction = int.dimred,
                         verbose = TRUE)
  }

}

# Re-join layers after integration
seu[["RNA"]] <- JoinLayers(seu[["RNA"]])

# Run UMAP for every integration method
int.umaps.plots <- list()
for (m in names(int.methods)[1:4]) {
  cat("\nRunning UMAP for", m, "integrated result...\n")
  int.dimred <- paste0("integrated.", m)
  umap.dimred <- paste0("umap.", m)
  seu <- RunUMAP(seu, dims = 1:30, reduction = int.dimred, reduction.name = umap.dimred)
  int.umaps.plots[[m]] <-  DimPlot(object = seu, reduction = umap.dimred, group.by = c("batch", "cell_type"), 
                                   combine = FALSE, label.size = 2)
}
```

<br>

<br>

---

<br>

<br>

## (4) Assess integration

_(7 min)_

<br>

```{r assess integration - plot, fig.width=16, fig.height=20}
## Assess integration by printing the plots using the "batch" and "cell_type" (ground-truth) labels
wrap_plots(c(int.umaps.plots$CCA, int.umaps.plots$RPCA, int.umaps.plots$Harmony, int.umaps.plots$FastMNN),
           ncol = 2, byrow = TRUE)
```

<br>

<br>

---

<br>

<br>

#### R packages used and respective versions

<br>

```{r References, message=FALSE, warning=FALSE, paged.print=FALSE}
## R packages and versions used in these analyses
sessionInfo()
```

<br>

<br>

---

<br>

<br>
